### DISCUSSION TITLE
Decorators typings between two pyright versions

### DISCUSSION THREAD

**Discussion by Aokigahara23** [microsoft/pyright#6425] [Q&A] [answered]
Hello community.

First of all thank you for the `pyright`, so far it is the best LSP for nvim i've found and used.

I'm having trouble understading what has changed between `1.1.334` -> `1.1.335`

Im using `prefect` framwork in my work alot. The core of it is built around so-called `flows` and `tasks` decorators [flow decorator API reference](https://docs.prefect.io/latest/api-ref/prefect/flows/#prefect.flows.flow) / [task decorator API reference](https://docs.prefect.io/latest/api-ref/prefect/tasks/#prefect.tasks.task). So, as an example i have this:

```python
from prefect import flow, task


@flow()
def my_flow() -> str:
    return my_task()

@task()
def my_task() -> str:
    return "Hello, world!"

result = my_flow()
```

In terms of typing, guys from prefect made a good job to make their API very comfortable to use, so for me those @decorators in this example are very transparent, so i will have `result` variable resolved to a `str` with no problem

But even in this case pyright LSP server returns me a `hover doc` over my_flow() function call as 
```text
pyright 1.1.334

(function) my_flow: Flow[(), str]
```
versus
```
pyright 1.1.335

(function) def my_flow() -> str
```

Here is a `flow`  [decorator overload stack](https://github.com/PrefectHQ/prefect/blob/main/src/prefect/flows.py#L1195)
This decorator returns a `Flow` object which has `__call__` [method overload](https://github.com/PrefectHQ/prefect/blob/main/src/prefect/flows.py#L1021)

So in case i provide `return_state` parameter to `true`, i should get `State` object as a `flow` return value, and i do with pyright==1.1.334

```python
# pyright 1.1.334

from prefect import flow, task


@flow()
def my_flow() -> str:
    return my_task()

@task()
def my_task() -> str:
    return "Hello, world!"

result = my_flow(return_state=True)

# (variable) result: State[str]
```

But with pyright 1.1.335 i have this
```python
# pyright 1.1.334

@flow()
def my_flow() -> str:
    return my_task()

@task()
def my_task() -> str:
    return "Hello, world!"

result = my_flow(return_state=True)     # 1. Pyright: No parameter named "return_state" [reportGeneralTypeIssues]

# (variable) result: Unknown      <- obvious because of the error above
```

Can you please tell me what am i doing wrong? Or maybe the latest changes require some additional actions from me, like changing the lsp settings. Cuz right now i'm using bare defaults and i'm okay with them

```lua
-- configure python server
lspconfig["pyright"].setup({
	capabilities = capabilities,
	on_attach = on_attach,
})
```

**Comment by erictraut** [comment, marked as answer]
The problem here is that the `prefect` library's class `Flow` has a class decorator `@prefectObjectRegistry.register_instances`. This decorator function is missing a return type annotation. This means pyright attempts to infer its return type. Its return type is based on the input parameter `type_: Type`. The type annotation `Type` is missing a type argument, so pyright assumes `Type[Unknown]` (where `Unknown` is an implied `Any`). The resulting return type for the decorator function is therefore `Type[Unknown]`.

Pyright typically honors class decorators, but it has heuristics where it ignores the decorator type under certain circumstances. The heuristics changed slightly between 1.1.334 and 1.1.335 due to a bug fix in another part of the code. In versions of pyright prior to 1.1.335, the decorator was ignored. Now it's not.

I don't consider this a bug in pyright, as I think it's reasonable for a type checker to apply the class decorator in this case. After all, the decorator function has a parameter type annotation, so it's not completely without type information.

The correct fix here is for the maintainers of prefect to properly annotate the function `register_instances`. It should look like this:

```python
    def register_instances(cls, type_: Type[T]) -> Type[T]:
```

Library maintainers might also be interested in using the "--verifytypes" feature of the pyright CLI to detect other places where their type information is incomplete. When I run `pyright --verifytypes prefect` on the latest published version of the library, here's the output:

```
Symbols exported by "prefect": 6664
  With known type: 3026
  With ambiguous type: 193
  With unknown type: 3445

Type completeness score: 45.4%
```

For a library that claims to provide type information (i.e. ships with a "py.typed" marker), 45.4% is a relatively low type completeness score. Providing full type information for a library of this size is not a small undertaking, and it's great to see that they've started down this road, but it looks like they still have quite a bit more work to provide full type information.

**Reply by Aokigahara23** [reply]
My appreciation for such a detailed answer, i will create an issue in [PrefectHQ](https://github.com/PrefectHQ/prefect) with a link to this thread. I hope the guys from their team will be interested in your opinion.
